### PR TITLE
Add XHR and Same-origin Policy

### PR DESCRIPTION
**Pull request by pavpanchekha**
This PR adds a section on `XMLHttpRequest` and also implements the Same-origin policy to restrict how the response can be used. Basic CORS stuff is added as an exercise. To make things flow better, I also reordered the text to do SOP then CSRF then XSS.

The XHR implementation is very very minimal / lazy, and I could be convinced it needs a few extra checks.

### CODE REVIEW

File: book/security.md [hunk 1]
```diff
@@ -405,184 +406,192 @@ them steal cookie values...
...
-the code in our guest book that outputs guest book entries:
+The easiest way for an attacker to steal your private data is to ask
+for it. Of course, there's no API in the browser for a website to ask
+for another website's cookies. But there _is_ an API to make requests
+to another website. It's called `XMLHttpRequest`.[^weird-name]
 
-``` {.python file=server indent=8}
-out += "<p>" + entry + " <i>by " + who + "</i></p>"
-```
+[^weird-name]: It's a weird name! Why is `XML` capitalized but not
```
Comment (chrishtr) [inline book/security.md:423]: Are you sure it can handle non-HTTP requests? I'm not sure that's correct.

Comment (chrishtr) [reply book/security.md:423]: Add: "The XML name comes from the historical origins of the API as part of the Outlook Web Access project at Microsoft (https://en.wikipedia.org/wiki/XMLHttpRequest)

File: book/security.md [hunk 2]
```diff
@@ -405,184 +406,192 @@ them steal cookie values...
...
+        full_url = resolve(url, self.tab.url)
+        headers, out = request(full_url, payload=body)
+        return out
 ```
 
-This is a good fix, and you should do it. But if you forget to encode
-any text anywhere---that's a security bug. So browsers provide
-additional layers of defense.
+With this new JavaScript API, a web page can make HTTP requests while
+the user interacts with it, making various "live", "AJAX" websites
```
Comment (chrishtr) [inline book/security.md:498]:
```suggestion
the user interacts with it, making more interactive websites
```

File: book/security.md [hunk 3]
```diff
@@ -405,184 +406,192 @@ them steal cookie values...
...
+possible! This API, and newer analogs like [`fetch`][mdn-fetch], are
+the basis for so many web page interactions that don't reload the
```
Comment (chrishtr) [inline book/security.md:500]:
```suggestion
the basis for many web page interactions that don't reload the
```

File: book/security.md [hunk 4]
```diff
@@ -405,184 +406,192 @@ them steal cookie values...
...
+page---think "liking" a post or hover previews or similar.
```
Comment (chrishtr) [inline book/security.md:501]:
```suggestion
page---think "liking" a post, hover previews or submitting a form without reloading the page.
```

File: book/security.md [hunk 5]
```diff
@@ -405,184 +406,192 @@ them steal cookie values...
...
-get a `<script>` added to the page, the browser would refuse to load
-and run that script.
+Same-origin Policy
+==================
 
-Let's implement support for this header. First, we'll need to extract
-and parse the `Content-Security-Policy` header:[^more-complex]
+However, new capabilities lead to new responsibilities. After all:
+any HTTP requests triggered by `XMLHttpRequest` will include cookies,
+which means they can potentially be used to steal or abuse cookies!
```
Comment (chrishtr) [inline book/security.md:512]:
```suggestion
which means they can potentially be used to steal or abuse private information.
```

File: book/security.md [hunk 6]
```diff
@@ -405,184 +406,192 @@ them steal cookie values...
...
+your account. But it also means that `XMLHttpRequest`s have access to
+private data, and thus need to protect it.
 
-[^more-complex]: In reality `Content-Security-Policy` can also list
-    scheme-generic URLs and even more complex sources like `'self'`.
-    This is a very limited CSP implementation.
+Let's imagine an attacker that wants to know your username on our
+guest book server. When you're logged in, the guest book includes your
+username in the "Hello, so and so" header, so it's enough for the
+attacker to read the guest book web page with your cookies.
```
Comment (chrishtr) [inline book/security.md:521]:
```suggestion
attacker to read the guest book web page with your cookies, in order to find out your name
and violate your privacy.
```

File: book/security.md [hunk 7]
```diff
@@ -405,184 +406,192 @@ them steal cookie values...
...
+page from reading the guest book web page content.
 
-``` {.python}
-class Tab:
-    def allowed_request(self, url):
-        if self.allowed_servers == None: return True
-        scheme_colon, _, host, _ = url.split("/", 3)
-        return scheme_colon + "//" + host in self.allowed_servers
-```
+The term for this is the "Same-origin policy". Basically, web pages
```
Comment (chrishtr) [inline book/security.md:546]:
```suggestion
This is achieved by enforcing a *same-origin policy*. Requests subject to this policy enforce that web pages
```

File: book/security.md [hunk 8]
```diff
@@ -405,184 +406,192 @@ them steal cookie values...
...
+can only make `XMLHttpRequests` for web pages on the same
```
Comment (chrishtr) [inline book/security.md:547]:
```suggestion
 only make `XMLHttpRequests` for web pages on the same
```

File: book/security.md (hunk 7 above)
Comment (chrishtr) [reply book/security.md:546]: Link: https://developer.mozilla.org/en-US/docs/Web/Security/Same-origin_policy

File: book/security.md [hunk 9]
```diff
@@ -847,23 +1044,34 @@ The server is much simpler, but has also grown since last chapter:
...
-This is a good default,[^23] but is sometimes annoying. A server can
-set an `origin` parameter in the `Set-Cookie` header to strip off some
-subdomains from the cookie origin. Implement this in your browser,
-making sure to send these generalized-origin cookies on any requests
-covered by the generalized origin.
+overwritten if the same cookie is set again with a later date. On the
+server side, save the same expiration dates in the `SESSIONS` variable
+and use it to delete old sessions to save memory.
+
+*CORS*: Web servers can *opt in* to allowing cross-origin
```
Comment (chrishtr) [inline book/security.md:1065]:
```suggestion
*CORS*: Web servers can [*opt in*](https://developer.mozilla.org/en-US/docs/Web/HTTP/CORS) to allowing cross-origin
```

File: book/security.md [hunk 10]
```diff
@@ -847,23 +1045,34 @@ The server is much simpler, but has also grown since last chapter:
...
-making sure to send these generalized-origin cookies on any requests
-covered by the generalized origin.
+overwritten if the same cookie is set again with a later date. On the
+server side, save the same expiration dates in the `SESSIONS` variable
+and use it to delete old sessions to save memory.
+
+*CORS*: Web servers can [*opt in*](https://developer.mozilla.org/en-US/docs/Web/HTTP/CORS) to allowing cross-origin
+`XMLHttpRequest`s using the `Access-Control-Allow-Origin` header. The
+way it works is that on cross-origin HTTP requests, the web browser
+sends an `Origin` header with the origin of the requesting site. By
```
Comment (chrishtr) [inline book/security.md:1069]: Also explain that such requests to the server will contain a cookie from the origin of the get request URL.

File: book/security.md [hunk 11]
```diff
@@ -847,23 +1045,34 @@ The server is much simpler, but has also grown since last chapter:
...
+default, the browser then throws away the response to prevent private
+data from leaking. But if the server sends the
+`Access-Control-Allow-Origin` header, and its value is either the
+requesting origin or the special `*` value, the browser instead makes
+the output available to the script. All requests made by your browser
+will be what the CORS standard calls "simple requests".
 
 *Referer*: When your browser visits a web page, or when it loads a CSS
 or JavaScript file, it sends a `Referer` header[^24] containing the
```
Comment (chrishtr) [inline book/security.md:1078]:
```suggestion
or JavaScript file, it sends a `Referer` header[^referer] containing the
```

File: book/security.md [hunk 12]
```diff
@@ -875,7 +1084,4 @@ they don't want revealed to other websites, so browsers support a
 `same-origin` (only do so if navigating to another page on the same
 origin). Implement those two values for `Referer-Policy`.
 
-[^23]: Should `microsoft.com` read `google.com`? What about
-    `microsoft.co.uk` and `google.co.uk`?
-
 [^24]: Yep, spelled that way.
```
Comment (chrishtr) [inline book/security.md:1087]:
```suggestion
[^referer]: Yep, spelled that way.
```

File: book/security.md (hunk 1 above)
Comment (pavpanchekha) [reply book/security.md:423]: I did edit this footnote to include the history.

I'll remove the HTTP comment, I agree it's a stretch.

File: book/security.md (hunk 11 above)
Comment (pavpanchekha) [reply book/security.md:1078]: Thanks. Yep, the initial footnotes were auto-exported from Org-mode, and it's good to use descriptive names as we edit.

File: book/security.md [hunk 13]
```diff
@@ -385,10 +387,9 @@ need to make sure cookie data is safe from malicious actors. After
 all: if someone stole your `token` cookie, they could copy it into
 their browser, and the server would think they are you.
 
-Our browser already prevents *other servers* from seeing your cookie
-values, because it stores cookies per-host.[^11][^12][^13][^14] But
-attackers might be able to get *your server* or *your browser* to help
-them steal cookie values...
+Our browser must prevent one servers from seeing *another server's*
+cookie values.[^11][^12][^13][^14] But attackers might be able to get
```
Comment (chrishtr) [inline book/security.md:391]: Optional: rename these footnotes if you're up for it.

**Pull request merged**